TypeCobol.LanguageServer/FileWatchers/CopyWatcher.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+
+namespace TypeCobol.LanguageServer
+{
+    public class CopyWatcher : IDisposable
+    {
+        private Workspace _TypeCobolWorkSpace;
+        private List<FileSystemWatcher> fileWatchers = new List<FileSystemWatcher>();
+
+        public CopyWatcher(Workspace workspace)
+        {
+            _TypeCobolWorkSpace = workspace;
+        }
+
+        public void SetDirectoryWatcher(string directoryPath)
+        {
+            //Initialize File Watcher
+            var watcher = new FileSystemWatcher
+            {
+                Path = directoryPath,
+                NotifyFilter =
+                    NotifyFilters.LastAccess | NotifyFilters.LastWrite | NotifyFilters.FileName |
+                    NotifyFilters.DirectoryName
+            };
+            watcher.Filter = "*.cpy";
+
+            // Add event handlers.
+            watcher.Changed += OnChanged;
+            //No need to watch COPY creation
+            //If a copy is missing it'll be loaded once a parser refresh is called
+            //watcher.Created += OnChanged; 
+            watcher.Deleted += OnChanged;
+            watcher.Renamed += OnChanged;
+
+            //Start Watching files
+            watcher.EnableRaisingEvents = true;
+
+            fileWatchers.Add(watcher);
+        }
+
+        private void OnChanged(object sender, FileSystemEventArgs e)
+        {
+            var directory= new FileInfo(e.FullPath).Directory;
+            if (File.Exists(directory.FullName + Path.DirectorySeparatorChar + "~.lock"))
+                return;
+
+            _TypeCobolWorkSpace.CompilationProject.ClearImportedCompilationDocumentsCache();
+
+            Action refreshAction = () => { _TypeCobolWorkSpace.RefreshOpenedFiles(); };
+            lock (_TypeCobolWorkSpace.MessagesActionsQueue)
+            {
+                if (_TypeCobolWorkSpace.MessagesActionsQueue.All(mw => mw.Action != refreshAction))
```

Comment:
The comparison : `mw.Action != refreshAction` doesn't works because `refreshAction` is instanciated each time  `OnChanged` method is called.

As there also a refreshAction created in DependenciesFileWatcher class, can you ensure that there is only one  refreshAction either from DependenciesFileWatcher or CopyWatcher in the MessagesActionsQueue ?

Reply:
It's done